File: secator/tasks/x8.py
from urllib.parse import urlparse, urlunparse

from secator.decorators import task
from secator.definitions import (URL, PROXY, DATA, WORDLIST, RETRIES, OPT_NOT_SUPPORTED, USER_AGENT, THREADS, DELAY, TIMEOUT, RATE_LIMIT, METHOD, HEADER, FOLLOW_REDIRECT, FILTER_CODES, FILTER_REGEX, FILTER_SIZE, FILTER_WORDS, MATCH_CODES, MATCH_REGEX, MATCH_SIZE, MATCH_WORDS, DEPTH)  # noqa: E501
from secator.output_types import Url, Tag
from secator.serializers import JSONSerializer
from secator.tasks._categories import HttpParamsFuzzer


@task()
class x8(HttpParamsFuzzer):
	"""Hidden parameters discovery suite written in Rust."""
	cmd = 'x8'
	input_types = [URL]
	output_types = [Url, Tag]
	tags = ['url', 'fuzz', 'params']
	file_flag = '-u'
	input_flag = '-u'
	json_flag = '-O json'
	opt_prefix = '-'
	version_flag = '-V'
	opt_key_map = {
		DATA: '--body',
		USER_AGENT: OPT_NOT_SUPPORTED,
		THREADS: 'c',
		DEPTH: OPT_NOT_SUPPORTED,
		DELAY: '--delay',
		TIMEOUT: '--timeout',
		PROXY: 'x',
		METHOD: '--method',
		WORDLIST: 'w',
		FILTER_CODES: OPT_NOT_SUPPORTED,
		FILTER_REGEX: OPT_NOT_SUPPORTED,
		FILTER_SIZE: OPT_NOT_SUPPORTED,
		FILTER_WORDS: OPT_NOT_SUPPORTED,
		MATCH_CODES: OPT_NOT_SUPPORTED,
		MATCH_REGEX: OPT_NOT_SUPPORTED,
		MATCH_SIZE: OPT_NOT_SUPPORTED,
		MATCH_WORDS: OPT_NOT_SUPPORTED,
		HEADER: OPT_NOT_SUPPORTED,
		RETRIES: OPT_NOT_SUPPORTED,
		# HEADER: 'H',
		RATE_LIMIT: OPT_NOT_SUPPORTED,
		FOLLOW_REDIRECT: '--follow-redirects',
	}
	opt_value_map = {
		HEADER: lambda headers: ';'.join(headers.split(';;'))
	}
	item_loaders = [JSONSerializer()]
	install_pre_cmd = {
		'apk': ['build-base', 'pkgconf', 'libssl3', 'libcrypto3', 'openssl-dev'],
		'apt': ['build-essential', 'pkg-config', 'libssl-dev'],
		'pacman': ['base-devel', 'pkg-config', 'openssl'],
		'zypper': ['gcc', 'pkg-config', 'libopenssl-devel'],
		'*': ['gcc', 'pkg-config', 'openssl-devel'],
	}
	install_version = '4.3.0'
	install_cmd = 'cargo install x8@[install_version] --force'
	install_github_bin = False  # TODO: enable this once https://github.com/Sh1Yo/x8/issues/65 is fixed
	# install_github_version_prefix = 'v'
	# install_ignore_bin = ['alpine', 'ubuntu']
	github_handle = 'Sh1Yo/x8'
	proxychains = False
	proxy_socks5 = False
	proxy_http = True
	profile = 'io'

	@staticmethod
	def on_init(self):
		self.urls = []
		self.request_headers = {}
		for k, v in self.get_opt_value(HEADER, preprocess=True).items():
			self.request_headers[k] = v

	@staticmethod
	def on_json_loaded(self, item):
		url = item['url']
		parsed_url = urlparse(url)
		url_without_param = urlunparse(parsed_url._replace(query=''))

		if url not in self.urls:
			self.urls.append(url)
			yield Url(
				url=url,
				host=parsed_url.hostname,
				method=item['method'],
				status_code=item['status'],
				content_length=item['size'],
				request_headers=self.request_headers,
			)

		for param in item.get('found_params', []):
			extra_data = {k: v for k, v in param.items() if k != 'name'}
			extra_data['value'] = param['value']
			extra_data['url'] = url
			yield Tag(
				category='info',
				name='url_param',
				match=url_without_param,
				value=param['name'],
				extra_data=extra_data
			)

File: rust/crates/secator-tasks/src/x8.rs
//! x8 — hidden parameter discovery (Python `secator/tasks/x8.py`).
//!
//! Reads URLs (one per input), discovers query parameters that change the
//! response, emits a `Url(...)` per visited target and one
//! `Tag(category=info, name=url_param, value=<param>, match=<url_without_param>)`
//! per discovered parameter.

use secator_model::{Map, OutputItem, Tag, Url};
use secator_options::{FileMode, InputWiring, KeyMap, OptSchema, OptSpec, OptType, SingleMode};
use secator_runner::{empty_output_maps, HookCtx, HookRegistry, ItemLoader, ValidatorRegistry};
use serde_json::Value;
use url::Url as UrlParser;

use crate::TaskSpec;

pub static SPEC: TaskSpec = TaskSpec {
    name: "x8",
    description: "Hidden HTTP parameter discovery (Rust).",
    cmd: "x8",
    input_types: &["url"],
    output_types: &["url", "tag"],
    tags: &["url", "fuzz", "params"],
    json_flag: Some("-O json"),
    input_wiring: InputWiring { single: SingleMode::Flag("-u"), file: FileMode::Flag("-u") },
    item_loaders: &[ItemLoader::Json],
    input_chunk_size: 1,
    on_json_loaded: Some(on_json_loaded),
    on_regex_loaded: None,
    output_maps: empty_output_maps,
    discriminator: None,
    hooks: HookRegistry::EMPTY,
    validators: ValidatorRegistry::EMPTY,
    schema: build_schema,
    install: secator_runner::InstallSpec {
        version: Some("v4.3.1"),
        cmd: Some("cargo install --git https://github.com/freelabz/x8 --tag [install_version] x8 --force"),
        github_handle: Some("freelabz/x8"),
        cmd_pre: &[
            ("apk", &["build-base", "pkgconf", "libssl3", "libcrypto3", "openssl-dev"]),
            ("apt", &["build-essential", "pkg-config", "libssl-dev"]),
            ("pacman", &["base-devel", "pkg-config", "openssl"]),
            ("zypper", &["gcc", "pkg-config", "libopenssl-devel"]),
            ("*", &["gcc", "pkg-config", "openssl-devel"]),
        ],
        ..secator_runner::InstallSpec::EMPTY
    },
    proxy_caps: secator_runner::ProxyCaps::NONE,
    encoding: "utf-8",
    ignore_return_code: false,
    requires_sudo: false,
};

fn on_json_loaded(ctx: &mut HookCtx, item: Map) -> Vec<OutputItem> {
    let url_str = match item.get("url").and_then(|v| v.as_str()) {
        Some(s) => s.to_string(),
        None => return Vec::new(),
    };
    let parsed = match UrlParser::parse(&url_str) {
        Ok(u) => u,
        Err(_) => return Vec::new(),
    };
    let url_without_param = {
        let mut b = parsed.clone();
        b.set_query(None);
        b.to_string()
    };
    let mut out: Vec<OutputItem> = Vec::new();
    // De-dup Url emission per (target) — track in ctx.state.
    let seen_key = format!("x8:seen:{url_str}");
    if !ctx.state.contains_key(&seen_key) {
        ctx.state.insert(seen_key, "1".into());
        out.push(OutputItem::Url(Url {
            url: url_str.clone(),
            host: parsed.host_str().unwrap_or("").to_string(),
            method: item.get("method").and_then(|v| v.as_str()).unwrap_or("").to_string(),
            status_code: item.get("status").and_then(|v| v.as_i64()).unwrap_or(0),
            content_length: item.get("size").and_then(|v| v.as_i64()).unwrap_or(0),
            ..Default::default()
        }));
    }
    if let Some(params) = item.get("found_params").and_then(|v| v.as_array()) {
        for p in params {
            let name = p.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string();
            let mut extra: Map = Map::new();
            if let Some(o) = p.as_object() {
                for (k, v) in o {
                    if k != "name" {
                        extra.insert(k.clone(), v.clone());
                    }
                }
            }
            extra.insert("url".into(), Value::String(url_str.clone()));
            out.push(OutputItem::Tag(Tag {
                category: "info".into(),
                name: "url_param".into(),
                match_: url_without_param.clone(),
                value: name,
                extra_data: extra,
                ..Default::default()
            }));
        }
    }
    out
}

fn build_schema() -> OptSchema {
    let mut s = OptSchema { opt_prefix: "-", ..OptSchema::default() };
    s.meta_opts = crate::meta_opts::opts_http_base();
    crate::meta_opts::apply_config_defaults(&mut s.meta_opts);
    s.key_map.insert("data".into(), KeyMap::Flag("body".into()));
    s.key_map.insert("threads".into(), KeyMap::Flag("c".into()));
    s.key_map.insert("delay".into(), KeyMap::Flag("delay".into()));
    s.key_map.insert("timeout".into(), KeyMap::Flag("timeout".into()));
    s.key_map.insert("proxy".into(), KeyMap::Flag("x".into()));
    s.key_map.insert("method".into(), KeyMap::Flag("method".into()));
    s.key_map.insert("wordlist".into(), KeyMap::Flag("w".into()));
    s.key_map.insert("follow_redirect".into(), KeyMap::Flag("follow-redirects".into()));
    for k in [
        "user_agent", "depth", "filter_codes", "filter_regex",
        "filter_size", "filter_words", "match_codes", "match_regex",
        "match_size", "match_words", "header", "retries", "rate_limit",
    ] {
        s.key_map.insert(k.into(), KeyMap::NotSupported);
    }
    s.opts = vec![OptSpec {
        name: "wordlist",
        ty: OptType::Str,
        short: Some("w"),
        is_flag: false,
        default: None,
        help: "Wordlist of parameter names",
        internal: false,
        requires_sudo: false,
        shlex: true,
        pre_process: None,
        process: None,
    }];
    s
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn map_from(v: Value) -> Map { v.as_object().cloned().unwrap() }

    #[test]
    fn emits_url_and_tag_per_found_param() {
        let mut ctx = HookCtx::default();
        let out = on_json_loaded(
            &mut ctx,
            map_from(json!({
                "url": "https://example.com/?x=1",
                "method": "GET",
                "status": 200,
                "size": 1234,
                "found_params": [{"name": "debug", "value": "true"}]
            })),
        );
        let has_url = out.iter().any(|i| matches!(i, OutputItem::Url(_)));
        let tag = out.iter().find_map(|i| match i { OutputItem::Tag(t) => Some(t), _ => None }).unwrap();
        assert!(has_url);
        assert_eq!(tag.value, "debug");
    }
}
